test/unit/kubernetes-deploy/kubernetes_resource/deployment_test.rb

```diff
@@ -2,118 +2,277 @@
+
+      assert deploy.deploy_timed_out?
+      assert_equal "Timeout reason: Failed to progress\nLatest ReplicaSet: web-1", deploy.timeout_message.strip
+    end
+  end
+
+  def test_deploy_timed_out_based_on_progress_deadline_ignores_conditions_older_than_the_deploy
+    Timecop.freeze do
+      status = {
+        "replicas" => 3,
+        "conditions" => [{
+          "type" => "Progressing",
+          "status" => 'False',
+          "lastUpdateTime" => Time.now.utc - 10.seconds,
+          "reason" => "Failed to progress"
+        }]
+      }
+      deploy = build_synced_deployment(status: status, new_rs_status: { "replicas" => 1 }, max_unavail: 1)
+      deploy.kubectl.expects(:server_version).returns(Gem::Version.new("1.8")).at_least_once
 
-    deploy = KubernetesDeploy::Deployment.new(namespace: "foo", context: "", logger: logger, definition: rollout)
-    deploy.instance_variable_set(:@latest_rs, true)
+      deploy.deploy_started_at = nil # not started yet
+      refute deploy.deploy_timed_out?
 
-    assert_raises(RuntimeError) { deploy.deploy_succeeded? }
+      deploy.deploy_started_at = Time.now.utc - 4.seconds # 10s ago is before deploy started
+      refute deploy.deploy_timed_out?
+
+      deploy.deploy_started_at = Time.now.utc - 5.seconds # 10s ago is "equal" to deploy time (fudge for clock skew)
+      assert deploy.deploy_timed_out?
+    end
   end
 
-  def test_deploy_succeeded_raises_with_invalid_mix_of_annotation
-    rollout = {
-      'spec' => {
-        'strategy' => 'recreate'
-      },
-      'metadata' => {
-        'name' => 'fake',
-        'annotations' => { KubernetesDeploy::Deployment::REQUIRED_ROLLOUT_ANNOTATION => 'maxUnavailable' }
+  def test_deploy_timed_out_based_on_progress_deadline_accommodates_stale_conditions_bug_in_k8s_176_and_lower
+    Timecop.freeze do
+      status = {
+        "replicas" => 3,
+        "conditions" => [{
+          "type" => "Progressing",
+          "status" => 'False',
+          "lastUpdateTime" => Time.now.utc - 5.seconds,
+          "reason" => "Failed to progress"
+        }]
       }
-    }
+      deploy = build_synced_deployment(status: status, new_rs_status: { "replicas" => 1 }, max_unavail: 1)
+      deploy.deploy_started_at = Time.now.utc - 5.seconds # progress deadline of 10s has not elapsed
+      deploy.kubectl.expects(:server_version).returns(Gem::Version.new("1.7.6"))
 
-    kubectl_mock = Minitest::Mock.new
-    status_mock = Minitest::Mock.new
-    status_mock.expect :success?, true
-    kubectl_mock.expect(:run, [true, true, status_mock], [Object, Object, Object, Object, Object, Object])
-    deploy = KubernetesDeploy::Deployment.new(namespace: "", context: "", logger: logger, definition: rollout)
-    deploy.instance_variable_set(:@kubectl, kubectl_mock)
+      refute deploy.deploy_timed_out?
+    end
+  end
 
-    refute deploy.validate_definition
+  private
+
+  def build_synced_deployment(status:, new_rs_status:, rollout: nil, max_unavail:)
```

Comment:
I found `status` and `new_rs_status` confusing. How would you feel about `status` => `deployment_status` and `new_rs_status` => `rs_status`.

Does it make sense to default `max_unavail` to `false` and the other two params to `{}`?

Reply:
Good points. I'll try to rework this a bit.